script/version_up.py

```diff
@@ -0,0 +1,660 @@
+
+def _run_git(*cmd):
+    """Runs a git command on the current repository.
+
+  This function returs a proper utf8 string in success, otherwise it allows the
+  exception thrown by subprocess through.
+
+  e.g:
+    _run_git('add', '-u', '*.patch')
+  """
+    cmd = ['git'] + list(cmd)
+    return subprocess.check_output(cmd).strip().decode('utf-8')
+
+
+def _load_package_file(branch):
+    """Retrieves the json content of package.json for a given revision
+
+  Args:
+    branch:
+      A branch or hash to load the file from.
+  """
+    package = _run_git('show', f'{branch}:{PACKAGE_FILE}')
+    return json.loads(package)
+
+
+def _run_npm_command(*cmd):
+    """Runs an npm build command.
+
+  This function will run 'npm run' commands appended by any extra arguments are
+  passed into it.
+
+  e.g:
+    _run_npm_command('init')
+  """
+    cmd = ['npm', 'run'] + list(cmd)
+    process = subprocess.Popen(cmd, stdout=subprocess.PIPE, bufsize=-1)
+    while True:
+        line = process.stdout.readline()
+        if not line:
+            break
+        print(line.decode('utf-8').rstrip())
+
+    return process.wait() == 0
+
+
+def _update_pinslist_timestamp():
+    """Updates the pinslist timestamp in the input_file_parsers.cc file for the
+    version commit.
+    """
+    try:
+        with open(PINSLIST_TIMESTAMP_FILE, "r", encoding="utf-8") as file:
+            content = file.read()
+    except FileNotFoundError:
+        print(f"ERROR: File '{PINSLIST_TIMESTAMP_FILE}' not found. Aborting.")
+        sys.exit(1)
+
+    pattern = r"# Last updated:.*\nPinsListTimestamp\n[0-9]{10}\n"
+    match = re.search(pattern, content, flags=re.DOTALL)
+    if not match:
+        print('ERROR: Expected pattern for PinsListTimestamp block not found. '
+              'Aborting.')
+        sys.exit(1)
+
+    # Update the timestamp
+    timestamp = int(datetime.now().timestamp())
+    readable_timestamp = datetime.fromtimestamp(timestamp).strftime(
+        '%a %b %d %H:%M:%S %Y')
+    updated_content = re.sub(
+        pattern,
+        f"# Last updated: {readable_timestamp}\nPinsListTimestamp\n{timestamp}\n",
+        content,
+        flags=re.DOTALL,
+    )
+
+    # Write back to the file
+    with open(PINSLIST_TIMESTAMP_FILE, "w", encoding="utf-8") as file:
+        file.write(updated_content)
+
+    updated = _run_git('diff', PINSLIST_TIMESTAMP_FILE)
+    if updated == '':
+        raise Exception('Pinslist timestamp failed to update.')
+
+
+def _get_apply_patches_list():
+    """Retrieves the list of patches to be applied by running 
+    `npm run apply_patches`
+    """
+
+    process = subprocess.run([
+        'npm', 'run', 'apply_patches', '--', '--print-patch-failures-in-json'
+    ],
+                             capture_output=True,
+                             text=True,
+                             check=False)
+
+    # This is a regex to match the json output of the patches that failed to apply. No test was conduct regarding inner arrays.
```

Comment:
Not really clear about this comment - it says that the pattern is to match failures, but then if it doesn't match anything we print 'No patches to apply'.

Reply:
This regex is merely used to check if JSON array list was produced because `--print-patch-failures-in-json` doesn't always produce a JSON summary, if nothing failed to apply.